File: tests/typesafety/python/config_parse_once.py
# expect: passes
# title: parse_argv checks the port once, and Config carries the proof
# improves: config_validate_everywhere
from dataclasses import dataclass


@dataclass(frozen=True)
class Config:
    port: int
    verbose: bool


def parse_argv(argv: list[str]) -> Config:
    settings: dict[str, str] = {}
    for pair in argv:
        key, _, value = pair.partition("=")
        settings[key] = value
    port_text = settings.get("port")
    if port_text is None or not port_text.isdigit():
        raise ValueError("port missing or not a number")
    return Config(port=int(port_text), verbose=settings.get("verbose") == "true")


def connect(config: Config) -> str:
    return f"connecting on {config.port}"  # no check, and none needed


def report(config: Config) -> str:
    return f"listening on {config.port}"

File: tests/typesafety/typescript/config_parse_once.ts
// expect: passes

type Config = {
  readonly port: number;
  readonly verbose: boolean;
};

export function parseArgv(argv: string[]): Config {
  const settings = new Map<string, string>();
  for (const pair of argv) {
    const [key, value] = pair.split("=", 2);
    settings.set(key ?? "", value ?? "");
  }
  const portText = settings.get("port");
  if (portText === undefined || !/^\d+$/.test(portText)) {
    throw new Error("port missing or not a number");
  }
  const port = Number(portText);
  if (port < 1 || port > 65535) {
    throw new Error("port out of range");
  }
  const verboseText = settings.get("verbose") ?? "false";
  if (verboseText !== "true" && verboseText !== "false") {
    throw new Error("verbose must be true or false");
  }
  return { port, verbose: verboseText === "true" };
}

export function connect(config: Config): string {
  return `connecting on ${config.port}`;
}

export function report(config: Config): string {
  return `listening on ${config.port}`;
}
